### PR TITLE
Okx createOrder: unify 'stopPrice'

### PR DESCRIPTION
**Pull request by Dan-krm**
#12044 Added stop orders and take profit orders to Okx:
https://www.okx.com/docs-v5/en/#rest-api-trade-place-algo-order

Stop Order:
```
node examples/js/cli okx createOrder BTC/USDT:USDT conditional buy 1 40600 '{"slOrdPx":"40600","slTriggerPx":"40000","tdMode":"isolated","posSide":"long"}'

okx.createOrder (BTC/USDT:USDT, conditional, buy, 1, 40600, [object Object])
2022-03-12T07:36:37.317Z iteration 0 passed in 365 ms

{
  info: { algoId: '422779960817004546', sCode: '0', sMsg: '' },
  id: undefined,
  clientOrderId: undefined,
  timestamp: undefined,
  datetime: undefined,
  lastTradeTimestamp: undefined,
  symbol: 'BTC/USDT:USDT',
  type: 'conditional',
  timeInForce: undefined,
  postOnly: undefined,
  side: 'buy',
  price: undefined,
  stopPrice: undefined,
  average: undefined,
  cost: undefined,
  amount: undefined,
  filled: undefined,
  remaining: undefined,
  status: undefined,
  fee: undefined,
  trades: [],
  fees: []
}
```

Take Profit:
```
node examples/js/cli okx createOrder BTC/USDT:USDT conditional buy 1 30600 '{"tpOrdPx":"30600","tpTriggerPx":"30000","tdMode":"isolated","posSide":"long"}'

okx.createOrder (BTC/USDT:USDT, conditional, buy, 1, 30600, [object Object])
2022-03-12T07:38:49.432Z iteration 0 passed in 357 ms

{
  info: { algoId: '422780514968449026', sCode: '0', sMsg: '' },
  id: undefined,
  clientOrderId: undefined,
  timestamp: undefined,
  datetime: undefined,
  lastTradeTimestamp: undefined,
  symbol: 'BTC/USDT:USDT',
  type: 'conditional',
  timeInForce: undefined,
  postOnly: undefined,
  side: 'buy',
  price: undefined,
  stopPrice: undefined,
  average: undefined,
  cost: undefined,
  amount: undefined,
  filled: undefined,
  remaining: undefined,
  status: undefined,
  fee: undefined,
  trades: [],
  fees: []
}
```

### CODE REVIEW

**Comment by kroitor** [comment]
@Dan-krm @samgermain @frosty00 @ttodua @carlosmiei @DoctorSlimm @sc0Vu @pcriadoperez 

Let's review this aspect of the `stopPrice` unification to clarify it.
Related issue: https://github.com/ccxt/ccxt/issues/12044
PRs for reference:
- https://github.com/ccxt/ccxt/pull/12302
- https://github.com/ccxt/ccxt/pull/12316

First, there's two types/cases of "stops" and we have to differentiate between the two:

1. A standalone _conditional stop order_ aka _algo order_ – conditional orders can exist by themselves and can be placed separately from other orders.
2. A _stop loss order_ and/or a _take profit stop order_ – those orders can not exist by themselves and are usually tied to another order that opens a margin position (regardless of the market type, whether it is a contract-margin market or a spot-margin market).

This issue https://github.com/ccxt/ccxt/issues/12044 about the `stopPrice` unification is related to type/case (1) **ONLY**.

-------------------------------------------------------------------------------

The usage of type/case (1) is the following:

Suppose you want to place an order (b) if the price meets some condition. To do that, you can place a special order (a) called _"stop conditional"_ order.
A "stop conditional" order (a) is an order to place another "triggered" order (b). That triggered order (b) can be either an order to open a new contract-position, or it could be a spot-order, pretty much any kind of order. It does not care if you have a position or not. When the price condition is met, then the conditional stop order (a) will be triggered, and it will result in placing another triggered order (b) of type/amount/price that the user specifies. So, in general, to place a conditional stop order (a), the user has to supply:

1. conditional `stopPrice` – this is related to the conditional order (a) itself
2. triggered `type` (`'market'` or `'limit'`) – this is related to the order (b) that will be placed by the conditional order, when triggered
3. triggered `amount` – this is related to the order (b) that will be placed by the conditional order, when triggered
4. triggered `price` (if `'limit'`) – this is related to the order (b) that will be placed by the conditional order, when triggered

So, the `stopPrice` conditional order (a) type/case (1) is just a way of conditionally-placing another triggered order (b) if the market meets some condition. It will be triggered when that condition is met. If the desired condition never occurs again, then the conditional order (a) will never be triggered, it will wait until you cancel it. The order (b) won't be placed at all, if the condition is never met.

-------------------------------------------------------------------------------

The usage of type/case (2) is the following:

Suppose you want to open a contract-margin position or spot-margin position by placing an order (a).

You also want to close that position when your desired unrealized profit meets a certain level, in percentage or in absolute prices. At that point you want to realize or _"fix"_ your unrealized profits on that open position. You want to close a profitable position at some point with another order that would fix the profits on that position. That is called _"take profit stop"_ order. (b)

And you also want to close that position if the market goes against your expectation, and when your losses reach a certain level, you want to stop getting further losses, or _"fix" your losses. That is called a _"stop loss"_ order. (c)

In this case, the user wants to place order (a) and attach two stop-orders (b) and (c) to the resulting position that will be opened by order (a).
Therefore, the user needs to specify a set of params for all three orders:

1. opening order (a) amount
2. opening order (a) type (`'market'` or `'limit'`)
3. opening order (a) price (if `'limit'`)
4. closing order (b) `stopLossPrice` – this is optional
5. closing order (c) `takeProfitPrice` – this is optional

Note, that the amount is not specified for orders (b) and (c), cause those orders will usually close the entire `amount` of the position specified in order (a).
Since both orders (b) and (c) are optional, the user can skip any of the two if he wants to close that position manually.
Also, orders of type (b) and (c) can be added/attached later to an already existing open position.

So, stop-loss/take-profit orders cannot exist by themselves, cause their purpose is to fix losses and profits on existing positions (this works for both contract margin and spot margin).

There is no unification schema for type/case (2) at this time. It is yet to be designed. Probably with a pair of extra params like `stopLossPrice` and `takeProfitPrice` (as opposed to `stopPrice` in case (1)).

------------------------------------------------------------------------------

Once again, this issue https://github.com/ccxt/ccxt/issues/12044 about the `stopPrice` unification is related to type/case (1) **ONLY**.

**Comment by Dan-krm** [comment]
This PR is related to condition 1 that you highlighted. I'm a bit unsure of what changes I should make. The conditional orders use a different endpoint but the amount is already handled on line 1828.

Are you saying to use the unified stopPrice for the trigger price and to use the price argument for the actual order once the trigger is hit and if so what would be the best approach to include (params, 'slOrdPx') and the price argument in the same variable?

Also if it's a market order, would both the trigger price and the order price be equal to the price argument?

File: js/okx.js [hunk 1]
```diff
@@ -1868,13 +1868,21 @@ module.exports = class okx extends Exchange {
         const takeProfitTrigger = this.safeNumber (params, 'tpTriggerPx');
         params = this.omit (params, [ 'slOrdPx', 'stopPrice', 'tpOrdPx', 'tpTriggerPx', 'slTriggerPx' ]);
         if (stopPrice || stopTrigger) {
-            defaultMethod = 'privatePostTradeOrderAlgo';
-            request['slOrdPx'] = this.priceToPrecision (symbol, stopPrice);
-            request['slTriggerPx'] = this.priceToPrecision (symbol, stopTrigger);
+            if (stopPrice && stopTrigger) {
+                defaultMethod = 'privatePostTradeOrderAlgo';
+                request['slOrdPx'] = this.priceToPrecision (symbol, stopPrice);
```
Comment (kroitor) [inline js/okx.js:1873]: We should not use `stopPrice` for stop-loss orders, the `stopPrice` is used for stop-orders only:

- Stop orders use `stopPrice` << we should only unify this at this time.
- Stop-loss orders use `stopLossPrice` <<  this should not be mixed or confused with stopPrice
- Take-profit orders use `takeProfitPrice` << this should not be mixed or confused with stopPrice

**Comment by samgermain** [comment]
@Dan-krm Make sure that you can also fetch and cancel stoploss orders

**Comment by samgermain** [comment]
@Dan-krm Can you also fetch and cancel stoploss orders?

**Comment by Dan-krm** [comment]
> Can you also fetch and cancel stoploss orders?

Thanks for the reminder,
Okx fetchOpenOrders 'stopOrder' #12631
Okx cancelOrders 'stopOrder' #12632

**Pull request merged**